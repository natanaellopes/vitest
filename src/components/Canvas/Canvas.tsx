import { useState } from 'react';
import { Point } from './types';

export const Canvas = () => {
  const [pointList, setPointList] = useState<Point[]>([]);

  const handleCanvasClick = (event: React.MouseEvent<HTMLElement>) => {
    const isDuplicated = pointList.some(item => item.x === event.clientX && item.y === event.clientY);

    if (isDuplicated) {
      return;
    }

    setPointList(oldValue => ([...oldValue, {
      y: event.clientY,
      x: event.clientX
    }]));
  }

  return (
    <div id="canvas" onClick={handleCanvasClick} aria-hidden="true">
      {pointList.map(point => (
        <div key={`${point.x}-${point.y}`} className="points" style={{left: `${point.x - 7}px`, top: `${point.y - 7}px`}}></div>
      ))}
    </div>
  );
}