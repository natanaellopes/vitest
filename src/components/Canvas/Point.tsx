import React from "react";
import { Point as PointType } from "./types";

type PointProps = {
  point: PointType;
}

export const Point: React.FC<PointProps> = ({ point }) => (
  <div className="points" style={{left: `${point.x - 7}px`, top: `${point.y - 7}px`}}></div>
);