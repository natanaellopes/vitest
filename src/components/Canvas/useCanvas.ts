import { useState } from 'react';
import { Point } from './types';

export const useCanvas = () => {
  const [pointList, setPointList] = useState<Point[]>([]);

  const onCanvasClick = (event: React.MouseEvent<HTMLElement>) => {
    const isDuplicated = pointList.some(item => item.x === event.clientX && item.y === event.clientY);
  
    if (isDuplicated) {
      return;
    }
  
    setPointList(oldValue => ([...oldValue, {
      y: event.clientY,
      x: event.clientX
    }]));
  }

  return {
    pointList,
    onCanvasClick
  }
}

