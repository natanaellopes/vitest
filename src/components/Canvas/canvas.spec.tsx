import { it, describe, expect, vi } from 'vitest';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event'

import { Canvas } from "./Canvas";

const setup = () => render(<Canvas />);
const user = userEvent.setup()

describe('Canvas', () => {
  const getCanvas = (container: HTMLElement) => container.querySelector('#canvas') as HTMLElement;
  const getPoints = (canvas: HTMLElement) => canvas.getElementsByClassName('points');


  afterEach(() => {
    vi.clearAllMocks();
    vi.resetAllMocks();
  });

  it('O canvas não pode ter pontos quando inicia', () => {
    const { container } = setup();

    const canvas = getCanvas(container);

    const points = canvas?.getElementsByClassName('points');

    expect(points?.length).toBe(0);
  });

  it('Deve ser adicionado um ponto ao clicar no canvas', async () => {
    const { container } = setup();

    const canvas = getCanvas(container);

    await user.pointer({
      keys: '[MouseLeft]',
      target: canvas,
      coords: {
        x: 10,
        y: 10
      }
    });

    const points = getPoints(canvas);

    expect(points?.length).toBe(1);
    expect(points.item(0)).matchSnapshot();
  });

  it('Deve ser adicionados dois pontos ao clicar no canvas', async () => {
    const { container } = setup();

    const canvas = getCanvas(container);

    await user.pointer({
      keys: '[MouseLeft]',
      target: canvas,
      coords: {
        x: 10,
        y: 10
      }
    })

    await user.pointer({
      keys: '[MouseLeft]',
      target: canvas,
      coords: {
        x: 20,
        y: 20
      }
    })

    const points = getPoints(canvas);

    expect(points.length).toBe(2);

    expect(points.item(0)).matchSnapshot();
    expect(points.item(1)).matchSnapshot();
  });

  it('Não deve ser adicionado ponto repetido', async () => {
    const { container } = setup();

    const canvas = getCanvas(container);

    await user.pointer({
      keys: '[MouseLeft]',
      target: canvas,
      coords: {
        x: 10,
        y: 10
      }
    })

    await user.pointer({
      keys: '[MouseLeft]',
      target: canvas,
      coords: {
        x: 20,
        y: 20
      }
    })

    await user.pointer({
      keys: '[MouseLeft]',
      target: canvas,
      coords: {
        x: 20,
        y: 20
      }
    })

    const points = getPoints(canvas);

    expect(points.length).toBe(2);

    expect(points.item(2)).toBeNull();
  });

});