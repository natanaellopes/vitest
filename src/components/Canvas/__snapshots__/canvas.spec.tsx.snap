// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Canvas > Deve ser adicionado um ponto ao clicar no canvas 1`] = `
<div
  class="points"
  style="left: 3px; top: 3px;"
/>
`;

exports[`Canvas > Deve ser adicionados dois pontos ao clicar no canvas 1`] = `
<div
  class="points"
  style="left: 3px; top: 3px;"
/>
`;

exports[`Canvas > Deve ser adicionados dois pontos ao clicar no canvas 2`] = `
<div
  class="points"
  style="left: 13px; top: 13px;"
/>
`;
